### DISCUSSION TITLE
Implementing the `v:register` variable

### DISCUSSION THREAD

**Discussion by jphalip** [JetBrains/ideavim#1045] [Q&A]
In Vim (and IdeaVim), if you select some text then yank it, your cursor moves to the start of the selection after yanking. To change this behavior and keep the cursor in the original position, I've got this key map, which works in Vim:

```
vnoremap <expr> y 'm`"' . v:register . 'y``'
```

For example in Vim, if you do `V2j"ay`, then it stores the selected text in the `a` register and keeps your cursor in the original position.

This doesn't work in IdeaVim because the `v:register` variable isn't implemented yet.

So I tried to implement it in `VimVariableServiceBase.getVimVariable()`:

```
"register" -> {
    val registerChar = if (editor.carets().size == 1) injector.registerGroup.currentRegister else injector.registerGroup.getCurrentRegisterForMulticaret()
    VimString(registerChar.toString())
}
```

However, here `registerChar` doesn't contain the register set by the command. I also tried with `injector.registerGroup.lastRegisterChar`.

Can you tell what I might be missing?

**Comment by citizenmatt** [comment]
The problem is that you're looking at the register used for the currently executing command, rather than the register that should be used once the command has been completed. 

When we type `"adw`, the `a` register is selected before invoking `DeleteMotionAction`. When deleting text, `injector.registerGroup.lastRegisterChar` or `.currentRegister` will return this selected register and use it to store the deleted text. I _think_ we then reset this stored register once the command has finished (I'd need to confirm that).

But the mapping you provided is evaluated while the command is being built and before it's executed, so the register hasn't been set yet. You'll need to look at `CommandBuilder.register` to see what's currently been entered, if anything. If it's null, you should be able to fall back to `injector.registerGroup.defaultRegister` to get the required behaviour of `v:register` based on the `'clipboard'` option (again, would need testing).

You'll notice a TODO comment that wants to remove `CommandBuilder.register`. This is because there are very few reasons to be looking at the in-progress state of a command. Most of the time, callers don't want the in-progress state and should wait for the complete command. This is a legitimate reason for accessing in-progress state, so don't worry about the comment. Maybe we'll rename the accessor at some point, so it's more obvious that it's probably not what most people are looking for, and could still change depending on the rest of the entered command (like we did with `calculateCount0Snapshot`).

**Comment by jphalip** [comment]
Thanks again for the detailed explanation. I propose renaming the attribute `registerInEffect` and then here's an implementation:

```kotlin
"register" -> {
        // The name of the register in effect for the current normal mode
        // command (regardless of whether that command actually used a
        // register). Or for the currently executing normal mode mapping
        // (use this in custom commands that take a register).
        // If none is supplied it is the default register '"', unless
        // 'clipboard' contains "unnamed" or "unnamedplus", then it is
        // "*" or '+' ("unnamedplus" prevails).
        val register = KeyHandler.getInstance().keyHandlerState.commandBuilder.registerInEffect
          ?: when {
            injector.globalOptions().clipboard.contains("unnamedplus") -> '+'
            injector.globalOptions().clipboard.contains("unnamed") -> '*'
            else -> injector.registerGroup.defaultRegister
          }
        VimString(register.toString())
      }
```

I've based this on the documentation (https://neovim.io/doc/user/vvars.html#v%3Aregister) and Vim source code (https://github.com/vim/vim/blob/45e0704d9670c10bfaf2bb408d6a5cd639d23835/src/clipboard.c#L2210).

I'm going to look at adding some tests for it and then submit a PR. Let me know if you have any tips about the tests.

**Reply by citizenmatt** [reply]
Cool. Let's go with `registerSnapshot` or even a function `getSelectedRegisterSnapshot` to be consistent with `calculateCount0Snapshot`. The key thing is that it is a snapshot of the value, because further typing could change the register before the command is built - e.g. something like `"ad"bw` would return the `a` register if the `d` operator was mapped, and then `b` if the `w` motion was mapped (and the command will run with the `b` register selected).

Also, have you checked if `defaultRegister` already does the `'clipboard'` logic for you? I think we update the value when the option changes, so I think you might just be able to use `defaultRegister`.

**Reply by jphalip** [reply]
Oh right, yes the 'clipboard' logic seems to be handled here: https://github.com/JetBrains/ideavim/blob/365b58eb560ab4fbc7a0dc185f393d468682251c/vim-engine/src/main/kotlin/com/maddyhome/idea/vim/register/VimRegisterGroupBase.kt#L84-L98

**Comment by jphalip** [comment]
@citizenmatt So I've had to default to `currentRegister` instead of `defaultRegister` to handle a sequence like this: `"a:echo v:register`, which should return `a` because the register has already been recorded by the time you run the `echo` command.

Here's the PR: https://github.com/JetBrains/ideavim/pull/1046

I also have a few other open PRs if you have a chance to review them. Thanks again for your help!

**Reply by citizenmatt** [reply]
Nice catch! I didn't realise that `"a:echo v:register` would work in Vim 😁 

I'm only a contributor, so I can't really review your PRs, that's for @AlexPl292. I can take a look and leave a comment though.
